netreg: Initial import

authmillenon opened the pull request:
Depends on ~~#2397~~ (merged)  ~~#2440~~ (merged). Tests will follow when review is somewhat positive ;-)

haukepetersen commented on sys/include/net/ng_netreg.h:
```diff
@@ -0,0 +1,132 @@
+     *          E. g. protocol numbers / next header numbers in IPv4/IPv6,
+     *          ports in UDP/TCP, or similar.
+     */
+    uint32_t demux_ctx;
```
we put this in netapi and netdev to `uint16_t`...

authmillenon replied:
Question is: is this enough?

haukepetersen commented on sys/net/crosslayer/ng_netreg/ng_netreg.c:
```diff
@@ -0,0 +1,104 @@
+    return 0;
+}
+
+static ng_netreg_entry_t *_ng_netreg_getnext_internal(ng_netreg_entry_t *entry, kernel_pid_t proto_pid, uint32_t demux_ctx)
```
line length...

haukepetersen commented on sys/include/net/ng_netreg.h:
```diff
@@ -0,0 +1,132 @@
+ * @def NG_NETREG_SIZE
+ */
+#ifndef NG_NETREG_SIZE
+#define NG_NETREG_SIZE (NG_NETTYPE_NUMOF)
```
so the default setting is, that you can register one listener for each module?

authmillenon replied:
Yes, but with this macro adaption is easily possible

haukepetersen replied:
fine with me then.

haukepetersen commented on sys/net/crosslayer/ng_netreg/ng_netreg.c:
```diff
@@ -0,0 +1,104 @@
+
+    for (int i = 0; i < NG_NETREG_SIZE; i++) {
+        if (netreg_entries[i].next == NULL) {
+            entry = &(netreg_entries[i]);
```
hm, in the beginning all `netreg_entries[]` elements are null, so you always return `ENOMEM`?!

authmillenon replied:
You're right will fix this as soon as I can get to it.

haukepetersen commented on sys/include/net/ng_netreg.h:
```diff
@@ -0,0 +1,132 @@
+     *          ports in UDP/TCP, or similar.
+     */
+    uint32_t demux_ctx;
+    kernel_pid_t target_pid;    /**< The PID of the registering thread */
```
what do you need this PID for?

authmillenon replied:
To identify which layer you mean if you have to threads of the same implementation running (e. g. 6LoWPAN threads on top of 2  IEEE 802.15.4 radios).

haukepetersen commented on sys/include/net/ng_netreg.h:
```diff
@@ -0,0 +1,132 @@
+    /**
+     * @see clist_node_t::prev
+     */
+    struct ng_netreg_entry *prev;
```
maybe it makes sense to save some RAM by dropping this (and accept that we have a little code overhead when removing entries...).

OlegHahm replied:
Sounds sensible - since removing will probably occur rather seldom.

authmillenon replied:
That would be my sixths or seventh implementation of a single-linked list, I guess… when will the slist PR merged?

OlegHahm replied:
I'm not sure we need slist. Why not using utlist?

authmillenon replied:
Did not know this was merged already.

haukepetersen commented:
I am not quite convinced of this implementation, I think the lookup is still too expensive! (Compare to #2310)

OlegHahm commented:
What*s the relation to #2310?

OlegHahm commented on sys/net/crosslayer/ng_netreg/ng_netreg.c:
```diff
@@ -0,0 +1,104 @@
+    ng_netreg_entry_t *start = entry;
+
+    do {
+        if (entry->proto_pid == proto_pid && entry->demux_ctx == demux_ctx) {
```
parens

authmillenon commented:
@haukepetersen I was thinking about doing a lookup on a two-dimensional array (1 axis: type, other axis `demux_type % n`), but that seemed to expensive memory-wise. The lookup is in the very common case of 1:1 mappings nearly as fast as your implementation (I have just two comparisons more), since the list will be only one element long. Only for 1:n mappings there will be an iteration, but I don't expect more than two or maybe three protocols talking to the same lower layer in practice. On the network layer we have only IPv4 and IPv6, and we both already argued, that more network protocols are not very realistic. On the transport layer we currently have UDP and TCP and maybe we implement DCCP and SCTP someday. But it is not to expect that you need more than two of them for your application (either you use UDP and TCP together, or you use the benefits of either DCCP or SCTP to not having to do that ;-)). On application layer you might have more than that, but in general you will use the socket API as a "normal" user. And that is one thread only, too.

authmillenon commented:
@haukepetersen I just looked at your implementation again: How does your implementation caters to demultiplexing contexts?

authmillenon commented:
Utilizes utlist instead of clist now

haukepetersen commented:
>  I just looked at your implementation again: How does your implementation caters to demultiplexing contexts?

it doesn't. The context stuff was introduced after I did the PR...

authmillenon commented:
> The context stuff was introduced after I did the PR...

It was not (it was already in the [introductory PR of netapi](https://github.com/RIOT-OS/RIOT/pull/1448/files#diff-b3f87ab92401692831810482dfa1796cR319)), but this is not the point: With the demultiplexing context we either have to decide for a slower, but more memory efficient list implementation or a more memory-wasting, but faster, hash table implementation. A simple array-lookup is just not possible.

authmillenon commented:
A tree-based data structure would also be possible, but would need at least one extra pointer + the ROM size would be increasing since it would have a more complex code.

authmillenon commented:
Changed semantic, added tests and fixes following from them

haukepetersen commented:
Concerning the underlying data-structure: I think we could just provide more then one implementation in the future, one optimized for run-time and one for memory...

Otherwise this PR looks good to me: ACK

authmillenon commented:
Please wait for #2397 before merging.

authmillenon commented:
Squashed commits in scope of this PR

cgundogan commented on sys/net/crosslayer/ng_netreg/ng_netreg.c:
```diff
@@ -0,0 +1,149 @@
+#include "net/ng_nettype.h"
+#include "utlist.h"
+
+#define _INVALID_TYPE(type) (((type) == NG_NETTYPE_UNDEF) || ((type) >= NG_NETTYPE_NUMOF))
```
May this break, when someone puts a signed negative number in this macro? You could use `<=` instead of `==` to test for `NG_NETTYPE_UNDEF`

authmillenon replied:
On first: will do. On edit: don't understand what you're getting at.

authmillenon replied:
Ah, I now get your second comment: The implementation somewhat depends on `NG_NETTYPE_UNDEF` being 0. If it is anything else, it must be changed anyways. But since it is the only value that is defined in `ng_nettype.h` I don't expect it to change.

cgundogan commented on sys/net/crosslayer/ng_netreg/ng_netreg.c:
```diff
@@ -0,0 +1,149 @@
+    ng_netreg_entry_t *entry = NULL;
+
+    if (_INVALID_TYPE(type)) {
+        return EINVAL;
```
By using negative errno values we could reserve the positive range for other kind of feedbacks later?

authmillenon replied:
I don't see why... The API is well-defined enough.

haukepetersen replied:
I agree with @cgundogan, that a negative return value makes sense, and if it's only for consistency with other networking code (e.g. netdev).

authmillenon commented:
Had to rebase because of a weird merge error yesterday.
Addressed your comments

cgundogan commented:
> ERROR: The following new files generate doxygen warnings: sys/include/net/ng_netreg.h

authmillenon commented:
Fixed it, but since netapi is no part of this PR, travis will still complain.

OT: we need to find a common naming scheme for modules in doxygen. I assumed, they are called as the actual RIOT modules are, but @haukepetersen (or past-me?) called `netapi` e.g. `net_netapi`.

authmillenon commented:
I can however rebase to current master

authmillenon commented:
(but I wait for this until #2397 was merged ;-))

haukepetersen commented:
@authmillenon: I don't know if this is common understanding, but I always named modules using the `[core|sys|cpu|board|...]_[MODULE_NAME]` naming scheme for doxygen modules. I think this makes the names pretty save against collisions.

authmillenon commented:
Rebased to current master

authmillenon commented:
> @authmillenon: I don't know if this is common understanding, but I always named modules using the [core|sys|cpu|board|...]_[MODULE_NAME] naming scheme for doxygen modules. I think this makes the names pretty safe against collisions.

Wouldn't the name than be `sys_netapi`?

authmillenon commented:
See #2422

haukepetersen commented:
for my state -> yes. I actually never spend much thought on this, I put the `net_` prefix to the groups rather intuitively. But you are right, in a strict sense `sys` would do. Still wondering if `net_` is nicer as it groups all network stack stuff in one doxygen prefix?

authmillenon commented:
This PR needs #2440 too, and since it needed a rebase I rebased it on that.

Lotterleben commented on sys/include/net/ng_netreg.h:
```diff
@@ -0,0 +1,136 @@
+ * @return  -ENOMEM if no space is left in the registry.
+ * @return  -EINVAL if @p type was NG_NETTYPE_UNDEF or >= NG_NETTYPE_NUMOF
+ */
+int ng_netreg_register(ng_nettype_t type, uint16_t demux_ctx, kernel_pid_t pid);
```
why is ``demux_ctx`` not of type ``netapi_reg_demux_ctx_t ``?

Lotterleben commented on sys/include/net/ng_netreg.h:
```diff
@@ -0,0 +1,136 @@
+     *          E. g. protocol numbers / next header numbers in IPv4/IPv6,
+     *          ports in UDP/TCP, or similar.
+     */
+    uint16_t demux_ctx;
```
If I understood it correctly, the demultiplexing context works similar to the one from the netapi– why does the demultiplexing context in netapi gets its own typedef'ed type (``netapi_reg_demux_ctx_t ``) and here it's a ``uint16_t``?

authmillenon replied:
This type does not exist in the most current version of netapi, since we moved all of the registration to this module. `netapi_reg_demux_ctx_t` is quite a mouthful compared to a simple `uint16_t`, so I left it out ;)

Lotterleben commented on sys/include/net/ng_netreg.h:
```diff
@@ -0,0 +1,136 @@
+/**
+ * @brief   Registers a thread to the registry.
+ *
+ * @details The semanics are: Thread @p pid is interested in packets of
```
Missing t ^^

Lotterleben commented on sys/include/net/ng_netreg.h:
```diff
@@ -0,0 +1,136 @@
+int ng_netreg_register(ng_nettype_t type, uint16_t demux_ctx, kernel_pid_t pid);
+
+/**
+ * @brief   Removes a thread to the registry.
```
not “from the registry”?

Lotterleben commented on sys/net/crosslayer/ng_netreg/ng_netreg.c:
```diff
@@ -0,0 +1,155 @@
+        return -EINVAL;
+    }
+
+    for (int i = 0; i < NG_NETREG_SIZE; i++) {
```
To me, the purpose of this loop wasn't instantly clear, maybe it would make sense to add /* look for an unclaimed entry */ or something?

Lotterleben commented on sys/net/crosslayer/ng_netreg/ng_netreg.c:
```diff
@@ -0,0 +1,155 @@
+            /* prevent double registration */
+            LL_PREPEND(netreg[type - 1], entry);
+        }
+        else {
```
``/*entry is a duplicate, free again*/``

Lotterleben commented on sys/net/crosslayer/ng_netreg/ng_netreg.c:
```diff
@@ -0,0 +1,155 @@
+    }
+    else {
+        ng_netreg_entry_t *dup;
+
```
``/*Check if a similar entry already exists*/``

Lotterleben commented on sys/net/crosslayer/ng_netreg/ng_netreg.c:
```diff
@@ -0,0 +1,155 @@
+#define _INVALID_TYPE(type) (((type) <= NG_NETTYPE_UNDEF) || ((type) >= NG_NETTYPE_NUMOF))
+
+static ng_netreg_entry_t *netreg[NG_NETTYPE_NUMOF - 1]; /* leave out NG_NETTYPE_UNDEF */
+static ng_netreg_entry_t netreg_entries[NG_NETREG_SIZE];
```
``/*pool of all netreg entries*/``

Lotterleben commented on sys/net/crosslayer/ng_netreg/ng_netreg.c:
```diff
@@ -0,0 +1,155 @@
+
+#define _INVALID_TYPE(type) (((type) <= NG_NETTYPE_UNDEF) || ((type) >= NG_NETTYPE_NUMOF))
+
+static ng_netreg_entry_t *netreg[NG_NETTYPE_NUMOF - 1]; /* leave out NG_NETTYPE_UNDEF */
```
``/*pointers to entries of netreg_entries[], sorted by their type (ng_nettype_t). */``

Lotterleben commented on sys/net/crosslayer/ng_netreg/ng_netreg.c:
```diff
@@ -0,0 +1,155 @@
+        LL_SEARCH(netreg[type - 1], dup, entry, _netreg_entry_equal);
+
+        if (dup == NULL) {
+            /* prevent double registration */
```
This comment was confusing to me, because I first read it as “the call to LL_PREPEND prevents double registration”. Maybe it could be changed to something like ``/* entry is not a duplicate. Add to netreg */`` or something?

Lotterleben commented:
(I've had a bit of a hard time understanding ``ng_netreg_register()``, so I added some suggestions for explanatory comments)

authmillenon commented:
Addressed comments and moved from (only available during testing) reset- to init-based setting of the values, just in case someone decides KERNEL_PID_UNDEF can't be 0 anymore ;-)

Lotterleben commented:
:+1: I'll run the tests on a SAMR21 after lunch and if that goes through and Travis is happy, I'd ACK, too.

haukepetersen commented:
please squash.

authmillenon commented:
I'll wait till #2440 is merged.

haukepetersen commented:
sorry, overlooked that dependency...

haukepetersen commented:
now now now :-)

authmillenon commented:
Rebased and squashed. @Lotterleben asked if she can merge ;-)

haukepetersen commented:
let's just give Travis another chance to complain... @Lotterleben merge at will when Travis is happy!

Lotterleben commented:
Travis is happy, the tests on the SAMR21are happy... Imma merge this now :sparkles:

haukepetersen commented:
yeah!
